extra_data/tests/mockdata/agipd.py

```diff
@@ -0,0 +1,130 @@
+import numpy as np
+
+from .base import DeviceBase
+
+
+class AGIPDMDL(DeviceBase):
+    def __init__(
+        self, device_id,
+        rep_rate=True,
+        gain_setting=True,
+        integration_time=True
+    ):
+        super().__init__(device_id)
+        self.rep_rate = rep_rate
+        self.gain_setting = gain_setting
+        self.integration_time = integration_time
+
+        # A sample of some of the available keys.
+        self.control_keys = [
+            ('acquisitionTime', 'u2', ()),
+            ('bunchStructure.nPulses', 'u2', ()),
+            ('bunchStructure.firstPulse', 'u2', ()),
+            ('bunchStructure.methodIndex', 'u2', ()),
+            ('egsgat', 'u2', ()),
+            ('g1sgat', 'u2', ()),
+            ('g2sgat', 'u2', ()),
+            ('pcRowNr', 'u2', ()),
+            ('t0Delay', 'u8', ()),
+            ('ticolm', 'u2', ()),
+            ('vrfcds', 'u2', ()),
+        ]
+
+    def write_control(self, f):
+        super().write_control(f)
+
+        ctrl_grp = f'CONTROL/{self.device_id}'
+        run_grp = f'RUN/{self.device_id}'
+        for grp in [ctrl_grp, run_grp]:
+            length = self.ntrains if grp == ctrl_grp else 1
+
+            f[f"{grp}/setupr/value"] = np.array([32]*length)
```

Comment:
For the ones that are always set, I'd recommend putting them in the `control_keys` list above, so that's as complete as possible, and then filling out the values here, something like this:

```suggestion
            g = f[grp]
            g["setupr/value"][()] = 32
```

Reply:
(There are some similar examples for this in the modules `sidemic_camera` and `gec_camera`, although they're both for instrument data rather than control)

Reply:
yes, I kept getting errors with what I wrote if the properties are already in control_keys. This is why I removed it.

Thanks I will try the suggested code.

Reply:
Suggested code is in place. Thank you Thomas!